design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Long getLong(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a long.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Long getLong(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Double getDouble(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a double.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Double getDouble(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  Boolean getBoolean(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a boolean.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  Boolean getBoolean(String columnName);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnIndex index of column (1-indexed).
+   * @return column value.
+   */
+  BigDecimal getDecimal(int columnIndex);
+
+  /**
+   * Get the value for a particular column of the Row as a decimal.
+   *
+   * @param columnName name of column.
+   * @return column value.
+   */
+  BigDecimal getDecimal(String columnName);
+}
+```
+
+We considered representing column types in a more structured form (rather than plain strings) to accomodate complex/nested data types but felt this added complexity would not be helpful for most use cases. (This would also require a server-side change, so it's a fair bit of additional work.)
+
+For `getDecimal(...)` in the `Row` interface, rather than trying to parse the column type to extract the precision and scale, we will simply convert the value to a BigDecimal without explicitly specifying the precision and scale. We could also add an option for users to specify the scale and precision in the getter, if we think that would be useful.
+
+We could also add `getList(...)` and `getMap(...)` methods to the `Row` interface, but it's not clear to me how valuable this would be given that the nested data types would not be known.
+
+Tim also proposed switching from generic Java types (`Map<String, Object>`, `List<Object>`) to Vert.x types (`JsonObject`, `JsonArray`) as the latter is more type-safe and the client already has a dependency on Vert.x anyway. The downside, though, is that then apps that use the client would be required to depend on Vert.x as well.
+
+### Transient queries -- Non-streaming
+
+The `Client` interface will also provide the following methods for receiving the results of a transient query (push or pull) in a single batch (non-streaming),
```

Comment:
"push query" and "single batch (non-streaming)" contradicts each other. Also the headline is "Non-streaming"

Reply:
Disagree, the way the data is obtained is orthogonal to whether it's a push or a pull query.

Reply:
I am not sure. A push query returns a data stream while a pull query return a result set.

It does not make sense to "subscribe" to a result set IMHO.

Reply:
Again, I think you're confusing the meaning of "stream" in the ksqlDB server side mental model and whether results are "streamed" to the user. Like many words used in CS they unfortunately have multiple overloaded meanings used in different contexts.

Reply:
Again. Because of this, it would be helpful to chose the wording carefully to avoid confusion...

Reply:
The API is based on reactive streams. These are streams, calling them anything else would be strange imho.

Reply:
Cf https://github.com/confluentinc/ksql/pull/5236#discussion_r421191874